server/dive_server/crud_annotation.py

```diff
@@ -1,48 +1,211 @@
-        extra = 'forbid'
+        extra = 'ignore'
 
 
-def save_annotations(dsFolder: types.GirderModel, user: types.GirderUserModel, data: dict):
-    crud.verify_dataset(dsFolder)
-    track_dict = crud.getTrackData(crud.detections_file(dsFolder))
-    validated: AnnotationUpdateArgs = crud.get_validated_model(AnnotationUpdateArgs, **data)
+def save_annotations(
+    dsFolder: types.GirderModel,
+    upsert_list: Iterable[dict],
+    delete_list: Iterable[int],
+    user: types.GirderUserModel,
+    description="save",
+    overwrite=False,
+):
+    """
+    Annotations are lazy-deleted by marking their staleness property as true.
+    """
 
-    for track_id in validated.delete:
-        track_dict.pop(str(track_id), None)
-    for track in validated.upsert:
-        track_dict[str(track.trackId)] = track.dict(exclude_none=True)
+    datasetId = dsFolder['_id']
+    expire_operations = []  # Mark existing records as deleted
+    insert_operations = []  # Insert new records
+    changes = False
+    new_revision = get_last_revision(dsFolder) + 1
+    delete_annotation_update = {'$set': {REVISION_DELETED: new_revision}}
 
-    upserted_len = len(validated.delete)
-    deleted_len = len(validated.upsert)
+    if overwrite:
+        query = {DATASET: datasetId, REVISION_DELETED: {'$exists': False}}
+        AnnotationItem().collection.bulk_write(
+            [pymongo.UpdateMany(query, delete_annotation_update)]
+        )
+        changes = True
 
-    if upserted_len or deleted_len:
-        crud.saveTracks(dsFolder, track_dict, user)
+    for track_id in delete_list:
+        filter = {TRACKID: track_id, DATASET: datasetId, REVISION_DELETED: {'$exists': False}}
+        # UpdateMany for safety, UpdateOne would also work
+        expire_operations.append(pymongo.UpdateMany(filter, delete_annotation_update))
 
-    return {
-        "updated": upserted_len,
-        "deleted": deleted_len,
-    }
+    for newdict in upsert_list:
+        newdict.update({DATASET: datasetId, REVISION_CREATED: new_revision})
+        newdict.pop(REVISION_DELETED, None)
+        filter = {
+            TRACKID: newdict['trackId'],
+            DATASET: datasetId,
+            REVISION_DELETED: {'$exists': False},
+        }
+        if not overwrite:
+            # UpdateMany for safety, UpdateOne would also work
+            expire_operations.append(pymongo.UpdateMany(filter, delete_annotation_update))
+        insert_operations.append(pymongo.InsertOne(newdict))
+
+    # Ordered=false allows fast parallel writes
+    if len(expire_operations):
+        AnnotationItem().collection.bulk_write(expire_operations, ordered=False)
+        changes = True
+    if len(insert_operations):
+        AnnotationItem().collection.bulk_write(insert_operations, ordered=False)
+        changes = True
+
+    additions = len(insert_operations)
+    # Absolute value because expire_operations might be 0 if overwrite is true
+    deletions = abs(len(expire_operations) - additions)
```

Comment:
Just want to make sure we are good with this behavior.  On an Import of a json/csv file this will make the additions === deletions.  So if you import 50 tracks the additions and deletions will both be 50.  There is the additional note in the revision indicating that it is an import so that may be intended.

Reply:
No I messed it up again.